### DISCUSSION TITLE
The code that adds media types in MvcCoreBuilderExtensions.cs mean the response for JSON will ALWAYS will return `application/vnd.restful+json`

### DISCUSSION THREAD

**Discussion by VictorioBerra** [Dotnet-Boxed/Templates#333] [Q&A]
﻿**Which project template is the issue about?**
API
...

**Which version of the project template is the issue about?**

...

**Which version of Visual Studio or the dotnet SDK were you using?**
2.2.203
...

If you call the API with a client and you set Accept to `application/json` it is completely ignored and you will ALWAYS get `application/vnd.restful+json` back.

Is this desired? This does not seem like it should be that way.

https://github.com/Dotnet-Boxed/Templates/blob/master/Source/Content/ApiTemplate/Source/ApiTemplate/MvcCoreBuilderExtensions.cs#L94

**Comment by VictorioBerra** [comment]
The reason I think it should not be this way is that if you removed the second output formatter and just left `jsonOutputFormatterMediaTypes.Insert(0, ContentType.ProblemJson);` you **ALWAYS** get `"application/problem+json"` back. Even with a 200OK.

![image](https://user-images.githubusercontent.com/2934507/59388969-ce44b000-8d32-11e9-889a-99e6aa33227f.png)

I also tried setting Content-Type to no avail.

**Comment by RehanSaeed** [comment]
I'm not actually certain what the correct behaviour should be. If you request json and we send a specific subset of json, I'm not certain that is incorrect. Do you have some references detailing the correct behaviour?
Kind Regards,
Muhammad Rehan Saeed
RehanSaeed.com

**Comment by VictorioBerra** [comment]
@RehanSaeed I am not sure exactly I need to study "content negotiation" a little better. I assumed you should always get back what you said you accept. Getting problem json for a 200ok seems wrong.

**Comment by VictorioBerra** [comment]
Also, I think that since those are content subtypes (+json) you do not need to actually add them. https://github.com/khellang/Middleware/issues/54#issuecomment-501576334

**Comment by RehanSaeed** [comment]
I read @khellang's comment in https://github.com/khellang/Middleware/issues/54#issuecomment-501576334. If I've understood correctly, there isn't much else we can do. If so, can we close?

**Comment by VictorioBerra** [comment]
I vote that we remove all of this

https://github.com/Dotnet-Boxed/Templates/blob/master/Source/ApiTemplate/Source/ApiTemplate/MvcCoreBuilderExtensions.cs#L87-L94

**Comment by RehanSaeed** [comment]
Yeah, maybe. I feel like I should write a few unit tests for each scenario. Here is what I think currently happens or should happen. Not sure if any of this is correct. I'd love to find some time to find documentation or maybe we should ask a StackOverflow question and farm out that work.

- No Accept - Success returns: application/vnd.restful+json
- Accept: application/json - Success returns: application/vnd.restful+json
- Accept: application/vnd.restful+json - Success returns: application/vnd.restful+json

- No Accept - 400 returns: application/problem+json
- Accept: application/json - 400 returns: application/problem+json
- Accept: application/vnd.restful+json - 400 returns: application/problem+json

**Comment by VictorioBerra** [comment]
Yes that is what I would expect, maybe:

Accept: application/json - Success returns: application/json

I say that because i think we should be writing the least amount of extra boilerplate as possible. I would like to use whatever MVC Core uses out of the box.

**Comment by RehanSaeed** [comment]
That is the one I'm just not sure about and would need to see some evidence to back up.

Agree on writing the minimum boilerplate.

**Comment by RehanSaeed** [comment]
I messed with input and output formatters and wrote more integration tests for the ASP.NET Core 3.0 update. Please check out the netcore3 branch and comment.

**Comment by VictorioBerra** [comment]
The input/output formatters still look roughly the same right? They will still always return whatever you insert at the 0 index. I do like the test assertions for the problem details, I think that is helpful.

**Comment by RehanSaeed** [comment]
@VictorioBerra Did you manage to take a look at the tests?

One solution to the Swagger UI selecting an incorrect MIME type in the drop down is to add the following to the controllers. However, it's annoying to have to add this to all controllers. You could add it globally but sometimes you might want to use other MIME types, so I don't think thats a good idea.

```c#
[Produces("application/application/vnd.restful+json")]
[Produces("application/problem+json")]
```

This has been on my mind for a while. I need to experiment with it to see what to do. Feel free to do so yourself.

**Comment by VictorioBerra** [comment]
@RehanSaeed You just mean these content type checks you made on the tests right? https://github.com/Dotnet-Boxed/Templates/blob/master/Source/ApiTemplate/Tests/ApiTemplate.IntegrationTest/Controllers/CarsControllerTest.cs#L107

I think the way this works is fine for now. Id leave the `Input/OutputFormatterMediaTypes` stuff in the `MvcBuilderExtensions` as they are.

For the Swagger UI thing, that is definitely annoying.

I would also agree not to add those to every single controller, if you add globally is there a way to make that configurable in appsettings maybe? It seems like Swagger with force us to manually put it somewhere.

**Comment by RehanSaeed** [comment]
I wouldn't want to put it in globally as an MVC filter. That will cause problems when someone wants to respond with some other MIME type. Only place is on the `Controller`. I also forgot to mention that we'd need to also add `[Consumes()]` in a similar way.

**Comment by VictorioBerra** [comment]
I suppose its necessary then. Its too bad we cant add them to something like the new API conventions feature.

**Comment by RehanSaeed** [comment]
Yes, that might be one way to do it.

**Comment by RehanSaeed** [comment]
The MIME types in Swashbuckle are all over the place.

Adding `[Produces(...)]` and `[Consumes(...)]` all over the place or in the new API conventions is the only fix I can see currently. However, that is still not going to show the correct MIME type for each response example.

I think the ideal scenario is if `SwaggerResponseAttribute` could also take a MIME type:

```cs
[SwaggerResponse(StatusCodes.Status201Created, "The car was created.", typeof(Car), "application/vnd.restful+json")]
[SwaggerResponse(StatusCodes.Status400BadRequest, "The car is invalid.", typeof(ProblemDetails), "application/problem+json")]
```

Raised https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1691 to propose this.

**Comment by RehanSaeed** [comment]
Also wondering if I should stop using comments on controllers and switch completely to attributes.

**Comment by VictorioBerra** [comment]
I think the comments can provide more info than the attributes. Also, read this https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/816#issuecomment-405442154

**Comment by RehanSaeed** [comment]
Nice. You're one step ahead of me.

**Comment by RehanSaeed** [comment]
Still no answer in https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1691

**Comment by VictorioBerra** [comment]
I bet the change is a ton of work for them, especially since the UI stuff
is by a different team I think.

**Comment by RehanSaeed** [comment]
I'm not sure we're going to get an answer any time soon. On the other hand, I don't want to add `[Produces(...)]` and `[Consumes(...)]` to every controller action as that is kind of overkill and easy to get wrong when you consider exceptions.

**Comment by VictorioBerra** [comment]
Yeah, you could leverage the new API conventions but your point still stands. There would be exceptions. https://docs.microsoft.com/en-us/aspnet/core/web-api/advanced/conventions?view=aspnetcore-3.1
